### PR TITLE
BH-467: Allow pubsubClient to use native Google Service account mechanism

### PR DESCRIPTION
**Pull request by mr-prud**
…nism

**Description (for Contributor and Core Developer)**

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: src/Akeneo/Tool/Bundle/MessengerBundle/Transport/GooglePubSub/PubSubClientFactory.php [hunk 1]
```diff
@@ -26,8 +26,11 @@ public function __construct(string $keyFilePath)
 
     public function createPubSubClient(array $config): PubSubClient
     {
-        return new PubSubClient(array_merge([
-            'keyFilePath' => $this->keyFilePath
-        ], $config));
+        if (!empty($keyFilePath)) {
```
Comment (pyaillet) [inline src/Akeneo/Tool/Bundle/MessengerBundle/Transport/GooglePubSub/PubSubClientFactory.php:29]:
```suggestion
        if (!empty($this->keyFilePath)) {
```

**Review by pyaillet — CHANGES_REQUESTED** [review]

File: src/Akeneo/Tool/Bundle/MessengerBundle/Transport/GooglePubSub/PubSubClientFactory.php (hunk 1 above)
Comment (pchasle) [inline src/Akeneo/Tool/Bundle/MessengerBundle/Transport/GooglePubSub/PubSubClientFactory.php]: I don't get what this is fixing.

The `$config` here, should not include the `keyFilePath` currently.

So if the value was empty when the `PubSubFactory` was instantiated that means the google `PubSubClient` is constructed with `'keyFilePath' => null`.

And if that is the case, from what I can see in the google library `vendor/google/cloud-core/src/ClientTrait.php`  in the method `getKeyFile` it should just try to load it from the env var `GOOGLE_APPLICATION_CREDENTIALS`

Comment (pchasle) [reply src/Akeneo/Tool/Bundle/MessengerBundle/Transport/GooglePubSub/PubSubClientFactory.php]: If you did your test with this, it means the pim ignored the `keyFilePath` even if it was defined. And that wont be the case anymore.

Maybe it explain the underlying issue that the `keyFilePath` provided is wrong on flex ? or there should not be any.
And the env var GOOGLE_APPLICATION_CREDENTIALS as the right value.

Comment (pchasle) [reply src/Akeneo/Tool/Bundle/MessengerBundle/Transport/GooglePubSub/PubSubClientFactory.php]: The `keyFilePath` come from `SRNT_GOOGLE_APPLICATION_CREDENTIALS`

File: src/Akeneo/Tool/Bundle/MessengerBundle/Transport/GooglePubSub/PubSubClientFactory.php [hunk 2]
```diff
@@ -26,8 +26,11 @@ public function __construct(string $keyFilePath)
...
     public function createPubSubClient(array $config): PubSubClient
     {
-        return new PubSubClient(array_merge([
-            'keyFilePath' => $this->keyFilePath
-        ], $config));
+        if (!empty($this->keyFilePath)) {
+            return new PubSubClient(array_merge([
+                'keyFilePath' => $this->keyFilePath
+            ], $config));
+        }
```
Comment (tseho) [inline src/Akeneo/Tool/Bundle/MessengerBundle/Transport/GooglePubSub/PubSubClientFactory.php:33]:
```suggestion
        if (null !== $this->keyFilePath) {
            $config = array_merge([
                'keyFilePath' => $this->keyFilePath
            ], $config));
        }
```

it's a non-blocking suggestion.

**Review by tseho — APPROVED** [review]

**Review by willy-ahva — APPROVED** [review]

**Comment by mr-prud** [comment]
we will fix in the EE

**Pull request closed without merging**